src/types/type_param.rs

```diff
@@ -70,49 +57,38 @@ impl TypeArg {
-            for item in items {
-                check_type_arg(item, ty.as_ref())?;
+        (
+            TypeArg::Type(SimpleType::Classic(ClassicType::Hashable(_))),
+            TypeParam::Type(TypeTag::Hashable),
+        ) => Ok(()),
+        (TypeArg::Type(SimpleType::Classic(_)), TypeParam::Type(TypeTag::Classic)) => Ok(()),
+        (TypeArg::Type(_), TypeParam::Type(TypeTag::Simple)) => Ok(()),
+        (TypeArg::List(items), TypeParam::List(param)) => {
+            items.iter().try_for_each(|arg| check_type_arg(arg, param))
+        }
+        (TypeArg::Tuple(items), TypeParam::Tuple(types)) => {
+            let items_len = items.len();
+            let types_len = types.len();
+            if items.len() != types.len() {
+                Err(TypeArgError::WrongNumberTuple(items_len, types_len))
+            } else {
+                items
+                    .iter()
+                    .zip(types.iter())
+                    .try_for_each(|(arg, param)| check_type_arg(arg, param))
             }
-            Ok(())
         }
-        (TypeArg::Int(v), TypeParam::Value(HashableType::Int(width))) => {
-            check_int_fits_in_width(*v, *width).map_err(TypeArgError::Int)
+        (TypeArg::Usize(_), TypeParam::Usize) => Ok(()),
+        (TypeArg::Opaque(arg), TypeParam::Opaque(param)) => {
+            arg.check_custom_type(param)?;
+            Ok(())
         }
-        (TypeArg::String(_), TypeParam::Value(HashableType::String)) => Ok(()),
-        (arg, TypeParam::Value(HashableType::Container(ctr))) => match ctr {
-            Container::Opaque(_) => match arg {
-                TypeArg::CustomValue(_) => Ok(()), // Are there more checks we should do here?
-                _ => Err(TypeArgError::TypeMismatch(arg.clone(), param.clone())),
-            },
-            Container::List(elem) => check_type_arg(
-                arg,
-                &TypeParam::List(Box::new(TypeParam::Value((**elem).clone()))),
-            ),
-            Container::Map(_) => unimplemented!(),
-            Container::Tuple(_) => unimplemented!(),
-            Container::Sum(_) => unimplemented!(),
-            Container::Array(elem, sz) => {
-                let TypeArg::List(items) = arg else {return Err(TypeArgError::TypeMismatch(arg.clone(), param.clone()))};
-                if items.len() != *sz {
-                    return Err(TypeArgError::WrongNumber(items.len(), *sz));
-                }
-                check_type_arg(
-                    arg,
-                    &TypeParam::List(Box::new(TypeParam::Value((**elem).clone()))),
-                )
-            }
-            Container::Alias(n) => Err(TypeArgError::NoAliases(n.to_string())),
-        },
         _ => Err(TypeArgError::TypeMismatch(arg.clone(), param.clone())),
     }
 }
 
 /// Errors that can occur fitting a [TypeArg] into a [TypeParam]
-#[derive(Clone, Debug, PartialEq, Eq, Error)]
+#[derive(Clone, Debug, PartialEq, Error)]
```

Comment:
Hang on...no floats here. This is because of, what, CustomCheckFail ?

Reply:
yes, we don't seem to need to it so I removed.